ygot/struct_validation_map.go

```diff
@@ -314,3 +314,161 @@ func MergeJSON(a, b map[string]interface{}) (map[string]interface{}, error) {
+		case reflect.Ptr:
+			if err := copyPtrField(dstField, srcField); err != nil {
+				return err
+			}
+		case reflect.Interface:
+			if err := copyInterfaceField(dstField, srcField); err != nil {
+				return err
+			}
+		case reflect.Map:
+			if err := copyMapField(dstField, srcField); err != nil {
+				return err
+			}
+		case reflect.Slice:
+			if err := copySliceField(dstField, srcField); err != nil {
+				return err
+			}
+		default:
+			dstField.Set(srcField)
+		}
+	}
+	return nil
+}
+
+// copyPtrField copies srcField to dstField. srcField and dstField must be
+// reflect.Value structs which represent pointers.
+func copyPtrField(dstField, srcField reflect.Value) error {
+	if srcField.IsNil() || !srcField.IsValid() {
+		return nil
+	}
+
+	switch srcField.Elem().Kind() {
+	case reflect.Struct:
+		var d reflect.Value
+		d = reflect.New(srcField.Type().Elem())
+		if err := copyStruct(d.Elem(), srcField.Elem()); err != nil {
+			return err
+		}
+		dstField.Set(d)
+	default:
+		p := reflect.New(srcField.Type().Elem())
+		p.Elem().Set(srcField.Elem())
+		dstField.Set(p)
+	}
+
+	return nil
+}
+
+// copyInterfaceField copies srcField into dstField. Both srcField and dstField
+// are reflect.Value structs which contain an interface value.
+func copyInterfaceField(dstField, srcField reflect.Value) error {
+	if srcField.IsNil() || !srcField.IsValid() {
+		return nil
+	}
+
+	if srcField.Elem().Kind() != reflect.Ptr {
+		return fmt.Errorf("invalid interface type received: %T", srcField.Interface())
+	}
+
+	switch srcField.Elem().Elem().Kind() {
+	case reflect.Struct:
+		var d reflect.Value
+		d = reflect.New(srcField.Elem().Elem().Type())
+		if err := copyStruct(d.Elem(), srcField.Elem().Elem()); err != nil {
+			return err
+		}
+		dstField.Set(d)
+	default:
+		return fmt.Errorf("unknown type in interface, %s", srcField.Elem().Kind())
+	}
+	return nil
+}
+
+// copyMapField copies srcField into dstField. Both srcField and dstField are
+// reflect.Value structs which contain a map value.
+func copyMapField(dstField, srcField reflect.Value) error {
+	if srcField.Len() == 0 {
+		return nil
+	}
+	keys := srcField.MapKeys()
+	if k := srcField.MapIndex(keys[0]).Kind(); k != reflect.Ptr {
+		return fmt.Errorf("invalid map, got member type %s", k)
+	}
+
+	// TODO(robjs): When we move to go1.9+ only support in ygot, we can use the
+	// following to make a map of a particular size.
+	// nm := reflect.MakeMapWithSize(reflect.MapOf((keys[0]).Type(), srcField.MapIndex(keys[0]).Type()), srcField.Len())
+	nm := reflect.MakeMap(reflect.MapOf((keys[0]).Type(), srcField.MapIndex(keys[0]).Type()))
+	for _, k := range keys {
+		v := srcField.MapIndex(k)
+		d := reflect.New(v.Elem().Type())
+		if err := copyStruct(d.Elem(), v.Elem()); err != nil {
+			return err
+		}
+		nm.SetMapIndex(k, d)
+	}
+	dstField.Set(nm)
+	return nil
+}
+
+// copySliceField copies srcField into dstField. Both srcField and dstField are
+// reflect.Value structs which represent slices.
```

Comment:
Would you mind expanding the comment to cover what types of inputs are expected/supported and how the function behaves for the expected inputs (e.g. wrt overwriting vs appending)?
Without some explanation I'd assume that this function would recursively merge any duplicated elements in the slice but that doesn't seem to be the behavior here.

Reply:
I've added a TODO per the above on this. Previously it was an oversight not to have made these return errors. PTAL.